src/nm.c

```diff
@@ -962,49 +919,108 @@ netplan_netdef_write_nm(
-    if (s->len > len)
-        g_string_free_to_file(s, rootdir, "run/NetworkManager/conf.d/netplan.conf", NULL);
+     * auto-connect and interferes.
+     * Also, mark all devices managed by us explicitly, so it won't get in
+     * conflict with the system's udev rules that might ignore some devices
+     * in containers via usr/lib/udev/rules.d/85-nm-unmanaged-devices.rules */
+    GList* iter = np_state->netdefs_ordered;
+    while (iter) {
+        const NetplanNetDefinition* nd = iter->data;
+        const gchar* nm_type;
+        GString *tmp = NULL;
+        guint unmanaged = nd->backend == NETPLAN_BACKEND_NM ? 0 : 1;
+
+        /* Special case: manage or ignore any device of given type on empty "match: {}" stanza */
+        if (nd->has_match && !nd->match.driver && !nd->match.mac && !nd->match.original_name) {
+            nm_type = type_str(nd);
+            g_assert(nm_type);
+            g_string_append_printf(nm_conf, "[device-netplan.%s.%s]\nmatch-device=type:%s\n"
+                                            "managed=%d\n\n", netplan_def_type_name(nd->type),
+                                            nd->id, nm_type, !unmanaged);
+        }
+        /* Normal case: manage or ignore devices by specific udev rules */
+        else {
+            const gchar *prefix = "SUBSYSTEM==\"net\", ACTION==\"add|change|move\",";
+            const gchar *suffix = nd->backend == NETPLAN_BACKEND_NM ? " ENV{NM_UNMANAGED}=\"0\"\n" : " ENV{NM_UNMANAGED}=\"1\"\n";
+            g_string_append_printf(udev_rules, "# netplan: network.%s.%s (on NetworkManager %s)\n",
+                                   netplan_def_type_name(nd->type), nd->id,
+                                   unmanaged ? "deny-list" : "allow-list");
+            /* Match by explicit interface name, if possible */
+            if (nd->set_name) {
+                // simple case: explicit new interface name
+                g_string_append_printf(udev_rules, "%s ENV{ID_NET_NAME}==\"%s\",%s", prefix, nd->set_name, suffix);
+            } else if (!nd->has_match) {
+                // simple case: explicit netplan ID is interface name
+                g_string_append_printf(udev_rules, "%s ENV{ID_NET_NAME}==\"%s\",%s", prefix, nd->id, suffix);
+            }
+            /* Also, match by explicit (new) MAC, if available */
+            if (nd->set_mac) {
+                tmp = g_string_new(nd->set_mac);
+                g_string_append_printf(udev_rules, "%s ATTR{address}==\"%s\",%s", prefix, g_string_ascii_down(tmp)->str, suffix);
+                g_string_free(tmp, TRUE);
+            }
+            /* Finally, add a full match, using all rules & globs available
+             * from the "match" stanza (e.g. original_name/mac/drivers)
+             * This will match the "old" interface (i.e. original MAC and/or
+             * interface name) if it got changed */
+            if (nd->has_match && (nd->match.original_name || nd->match.mac || nd->match.driver)) {
```

Comment:
nitpick (non-blocking): this block could perhaps have been split off into its own function to make it a bit more bitesize ?

Reply:
Ugh.. I was a bit quick with merging and didn't remember this suggestion. It makes sense to encapsulate this into its own function. I will probably work on this in a follow-up commit.